talisker/metrics.py

```diff
@@ -112,3 +121,126 @@ def inc(self, amount=1, **labels):
             client = talisker.statsd.get_client()
             name = self.get_statsd_name(labels)
             client.incr(name, amount)
+
+
+histogram_archive = 'histogram_archive.db'
+counter_archive = 'counter_archive.db'
+
+
+def prometheus_cleanup_worker(pid):
+    """Clean up after a multiprocess worker has died."""
+    if prometheus_client is None:
+        return
+
+    try:
+        # just deletes delete gauge files
+        multiprocess.mark_process_dead(pid)
+        prom_dir = os.environ['prometheus_multiproc_dir']
+        pid_files = [
+            'histogram_{}.db'.format(pid),
+            'counter_{}.db'.format(pid),
+        ]
+        paths = [os.path.join(prom_dir, f) for f in pid_files]
+
+        # check at least one worker file exists
+        if not any(os.path.exists(path) for path in paths):
+            return
+
+        metrics = collect_metrics(prom_dir, *pid_files)
+
+        tmp_histogram = tempfile.NamedTemporaryFile(delete=False)
+        tmp_counter = tempfile.NamedTemporaryFile(delete=False)
+        write_metrics(metrics, tmp_histogram.name, tmp_counter.name)
+
+        os.rename(
+            tmp_histogram.name, os.path.join(prom_dir, histogram_archive))
+        os.rename(
+            tmp_counter.name, os.path.join(prom_dir, counter_archive))
+
+        for path in paths:
+            if os.path.exists(path):
+                os.unlink(path)
+
+    except Exception:
+        # we should never fail at cleaning up
+        logger.exception('failed to cleanup prometheus worker files')
+
+
+def collect_metrics(prom_dir, *files):
+    """Copy the files out of prom_dir into a separate dir, and collect them.
+
+    This aggregates all the metrics together into one set.
+    """
+    from prometheus_client import CollectorRegistry
+
+    all_files = [histogram_archive, counter_archive] + list(files)
+    tmp = tempfile.mkdtemp(prefix='prometheus_aggregate')
+    try:
+        for filename in all_files:
+            path = os.path.join(prom_dir, filename)
+            if os.path.exists(path):
+                shutil.copy(path, tmp)
+
+        collector = multiprocess.MultiProcessCollector(CollectorRegistry())
```

Comment:
Shouldn't `collector` receive the `tmp` path as a parameter? Or maybe I'm missing something.